File: src/moai_adk/core/git_lock_manager.py
"""
Git Lock Manager

Git 작업 동시 실행 방지를 위한 잠금 관리 시스템

@FEATURE:GIT-LOCK-001 - 동시 Git 작업 방지
@PERF:LOCK-100MS - 잠금 성능 최적화 (100ms 이내 응답)
@SEC:LOCK-MED - 잠금 파일 보안 강화
"""

import logging
import os
import time
from contextlib import contextmanager
from pathlib import Path
from typing import Generator

from .exceptions import GitLockedException
from .git_lock_acquirer import LockAcquirer
from .git_lock_file_handler import LockFileHandler
from .git_lock_validator import LockValidator

# 로깅 설정 (@TASK:LOG-001)
logger = logging.getLogger(__name__)


class GitLockManager:
    """Git 작업 동시 실행 방지를 위한 잠금 관리자

    .moai/locks/git.lock 파일을 사용하여 Git 작업의 동시 실행을 방지합니다.

    Features:
    - 모듈화된 아키텍처 (TRUST-U 원칙)
    - 성능 최적화된 잠금 메커니즘 (@PERF:LOCK-100MS)
    - 강화된 에러 처리 및 복구 (@SEC:LOCK-MED)
    - 구조화된 로깅 (@TASK:LOG-001)
    - 자동 정리 및 모니터링 (@FEATURE:AUTO-CLEANUP-001)
    """

    def __init__(self, project_dir: Path | None = None, lock_dir: str = ".moai/locks"):
        """Initialize GitLockManager

        Args:
            project_dir: 프로젝트 루트 디렉토리
            lock_dir: 잠금 파일이 저장될 디렉토리 경로 (project_dir 기준)
        """
        # 입력 검증 강화 (@SEC:LOCK-MED)
        if project_dir is None:
            project_dir = Path.cwd()
        elif isinstance(project_dir, str):
            project_dir = Path(project_dir).resolve()  # 절대 경로로 변환
        elif isinstance(project_dir, Path):
            project_dir = project_dir.resolve()
        else:
            raise ValueError(
                f"project_dir must be a Path or str type: {type(project_dir)}"
            )

        # Directory validation - allow non-existent directories during initialization
        if not project_dir.exists():
            logger.warning(
                f"Project directory will be created during installation: {project_dir}"
            )
            # Directory will be created by installer - defer validation

        self.project_dir = project_dir
        self.lock_dir = project_dir / lock_dir
        self.lock_file = self.lock_dir / "git.lock"

        # 모듈화된 컴포넌트 초기화
        self.file_handler = LockFileHandler(self.lock_file)
        self.validator = LockValidator()
        self.acquirer = LockAcquirer(self.file_handler, self.validator)

        logger.debug(
            f"GitLockManager 초기화: {self.project_dir}, 잠금파일: {self.lock_file}"
        )

    # API 호환성을 위한 위임 메서드들

    def is_locked(self) -> bool:
        """현재 잠금 상태 확인

        성능 최적화된 잠금 상태 검사 (@PERF:LOCK-100MS)

        Returns:
            잠금 파일이 존재하고 유효하면 True, 아니면 False
        """
        if not self.file_handler.lock_file_exists():
            return False

        # 잠금 파일 유효성 검증 (좀비 프로세스 확인)
        try:
            lock_info = self.file_handler.read_lock_info()
            if lock_info and self.validator.is_lock_valid(lock_info):
                return True
            else:
                # 무효한 잠금 파일 정리
                logger.info("무효한 잠금 파일 발견, 자동 정리 중")
                self.acquirer.cleanup_stale_lock()
                return False
        except Exception as e:
            logger.error(f"잠금 상태 확인 중 오류: {e}")
            return True  # 안전을 위해 잠금된 것으로 간주

    def release_lock(self):
        """잠금 파일 삭제

        강화된 잠금 해제 메커니즘 (@SEC:LOCK-MED)
        """
        self.acquirer.release_lock()

    def acquire_lock(self, wait: bool = True, timeout: int = 30):
        """잠금 획득

        컨텍스트 매니저로 사용하면 자동으로 잠금이 해제되고,
        직접 호출하면 검증만 수행합니다.

        Args:
            wait: 잠금 대기 여부
            timeout: 대기 시간 (초)

        Returns:
            contextmanager 또는 None

        Raises:
            GitLockedException: 잠금 획득에 실패한 경우
        """
        return self.acquirer.acquire_lock(wait, timeout)

    def acquire_lock_direct(self, wait: bool = True, timeout: int = 30) -> bool:
        """잠금 직접 획득 (컨텍스트 매니저 없이)

        Args:
            wait: 잠금 대기 여부
            timeout: 대기 시간 (초)

        Returns:
            bool: 잠금 획득 성공 여부

        Raises:
            GitLockedException: 잠금 획득에 실패한 경우
        """
        return self.acquirer.acquire_lock_direct(wait, timeout)

    def get_lock_status(self) -> dict:
        """잠금 상태 정보 반환

        모니터링 및 디버깅을 위한 상세 정보 제공

        Returns:
            잠금 상태 정보 딕셔너리
        """
        status = {
            "is_locked": self.is_locked(),
            "lock_file_exists": self.file_handler.lock_file_exists(),
            "lock_dir_exists": self.lock_dir.exists(),
            "lock_dir_writable": self.lock_dir.exists()
            and os.access(self.lock_dir, os.W_OK),
        }

        if status["lock_file_exists"]:
            lock_info = self.file_handler.read_lock_info()
            if lock_info:
                status.update(
                    {
                        "lock_info": lock_info,
                        "process_running": self.validator.is_process_running(
                            lock_info.get("pid", 0)
                        ),
                        "lock_age_seconds": self.validator.get_lock_age_seconds(lock_info),
                    }
                )

        return status

File: moai-adk-ts/src/core/git/git-lock-manager.ts
/**
 * @file Git Lock Manager for preventing concurrent Git operations
 * @author MoAI Team
 * @tags @FEATURE:GIT-LOCK-001 @REQ:CORE-SYSTEM-013
 */

import * as os from 'node:os';
import * as path from 'node:path';
import * as fs from 'fs-extra';
import {
  type GitLockContext,
  GitLockedException,
  type GitLockInfo,
  type GitLockStatus,
} from '../../types/git';

/**
 * GitLockManager class for preventing concurrent Git operations
 * Ports Python git_lock_manager.py with TypeScript improvements
 * @tags @TASK:GIT-LOCK-MANAGER-001
 */
export class GitLockManager {
  private readonly projectDir: string;
  private readonly lockDir: string;
  private readonly lockFile: string;
  private readonly maxLockAge: number = 300000; // 5 minutes
  private readonly pollInterval: number = 100; // 100ms

  constructor(projectDir?: string) {
    if (projectDir && typeof projectDir !== 'string') {
      throw new Error('Invalid project directory');
    }

    this.projectDir = projectDir || process.cwd();
    this.lockDir = path.join(this.projectDir, '.moai', 'locks');
    this.lockFile = path.join(this.lockDir, 'git.lock');
  }

  /**
   * Check if Git operations are currently locked
   * @returns True if locked, false otherwise
   * @tags @API:IS-LOCKED-001
   */
  public async isLocked(): Promise<boolean> {
    try {
      if (!(await fs.pathExists(this.lockFile))) {
        return false;
      }

      const lockInfo = await this.readLockInfo();
      if (!lockInfo) {
        await this.cleanupCorruptLock();
        return false;
      }

      // Check if process is still running
      if (!this.isProcessRunning(lockInfo.pid)) {
        console.log(`Cleaning up stale lock from process ${lockInfo.pid}`);
        await this.releaseLock();
        return false;
      }

      // Check if lock is too old
      const lockAge = Date.now() - lockInfo.timestamp;
      if (lockAge > this.maxLockAge) {
        console.log(`Cleaning up expired lock (age: ${lockAge}ms)`);
        await this.releaseLock();
        return false;
      }

      return true;
    } catch (error) {
      console.error('Error checking lock status:', error);
      // Clean up corrupt lock file
      await this.cleanupCorruptLock();
      return false;
    }
  }

  /**
   * Acquire lock with context manager pattern
   * @param wait Whether to wait for lock
   * @param timeout Timeout in seconds
   * @returns Lock context
   * @tags @API:ACQUIRE-LOCK-001
   */
  public async acquireLock(
    wait: boolean = true,
    timeout: number = 30
  ): Promise<GitLockContext> {
    const startTime = Date.now();
    const timeoutMs = timeout * 1000;

    while (true) {
      if (!(await this.isLocked())) {
        // Create lock
        const lockInfo = await this.createLock('unknown');
        return {
          lockInfo,
          acquired: new Date(),
          release: () => this.releaseLock(),
        };
      }

      if (!wait) {
        const existingLock = await this.readLockInfo();
        throw new GitLockedException(
          'Git operations are locked by another process',
          existingLock || undefined
        );
      }

      // Check timeout
      if (Date.now() - startTime > timeoutMs) {
        throw new GitLockedException(
          'Lock acquisition timeout',
          undefined,
          timeout
        );
      }

      // Wait before retrying
      await this.sleep(this.pollInterval);
    }
  }

  /**
   * Release the Git lock
   * @tags @API:RELEASE-LOCK-001
   */
  public async releaseLock(): Promise<void> {
    try {
      if (await fs.pathExists(this.lockFile)) {
        await fs.remove(this.lockFile);
        console.log('Git lock released');
      }
    } catch (error) {
      console.error('Error releasing lock:', error);
      // Don't throw - we want to be resilient
    }
  }

  /**
   * Execute operation with lock (context manager pattern)
   * @param operation Operation to execute
   * @param operationName Name of the operation for logging
   * @param wait Whether to wait for lock
   * @param timeout Timeout in seconds
   * @returns Operation result
   * @tags @API:WITH-LOCK-001
   */
  public async withLock<T>(
    operation: () => Promise<T>,
    operationName: string = 'unknown',
    wait: boolean = true,
    timeout: number = 30
  ): Promise<T> {
    const lockContext = await this.acquireLock(wait, timeout);

    try {
      console.log(`Executing Git operation: ${operationName}`);
      return await operation();
    } finally {
      await lockContext.release();
    }
  }

  /**
   * Get comprehensive lock status
   * @returns Lock status information
   * @tags @API:GET-LOCK-STATUS-001
   */
  public async getLockStatus(): Promise<GitLockStatus> {
    const lockFileExists = await fs.pathExists(this.lockFile);

    if (!lockFileExists) {
      return {
        isLocked: false,
        lockFileExists: false,
        lockInfo: null,
      };
    }

    const lockInfo = await this.readLockInfo();
    if (!lockInfo) {
      return {
        isLocked: false,
        lockFileExists: true,
        lockInfo: null,
      };
    }

    const processRunning = this.isProcessRunning(lockInfo.pid);
    const lockAgeSeconds = (Date.now() - lockInfo.timestamp) / 1000;

    return {
      isLocked: processRunning && lockAgeSeconds < this.maxLockAge / 1000,
      lockFileExists: true,
      lockInfo,
      processRunning,
      lockAgeSeconds,
    };
  }

  /**
   * Clean up stale locks
   * @tags @API:CLEANUP-STALE-LOCKS-001
   */
  public async cleanupStaleLocks(): Promise<void> {
    if (!(await fs.pathExists(this.lockFile))) {
      return;
    }

    const lockInfo = await this.readLockInfo();
    if (!lockInfo) {
      await this.cleanupCorruptLock();
      return;
    }

    // Check if process is still running
    if (!this.isProcessRunning(lockInfo.pid)) {
      console.log(`Cleaning up stale lock from dead process ${lockInfo.pid}`);
      await this.releaseLock();
      return;
    }

    // Check if lock is too old
    const lockAge = Date.now() - lockInfo.timestamp;
    if (lockAge > this.maxLockAge) {
      console.log(`Cleaning up expired lock (age: ${lockAge}ms)`);
      await this.releaseLock();
      return;
    }

    console.log('Lock is active and valid');
  }

  // Private helper methods

  /**
   * Create lock file with current process information
   */
  private async createLock(
    operation: string = 'unknown'
  ): Promise<GitLockInfo> {
    await fs.ensureDir(this.lockDir);

    const lockInfo: GitLockInfo = {
      pid: process.pid,
      timestamp: Date.now(),
      operation,
      user: os.userInfo().username,
      hostname: os.hostname(),
      workingDir: this.projectDir,
    };

    await fs.writeJson(this.lockFile, lockInfo, { spaces: 2 });
    console.log(
      `Git lock acquired by process ${lockInfo.pid} for operation: ${operation}`
    );

    return lockInfo;
  }

  /**
   * Read lock information from file
   */
  private async readLockInfo(): Promise<GitLockInfo | null> {
    try {
      if (!(await fs.pathExists(this.lockFile))) {
        return null;
      }

      const lockInfo = (await fs.readJson(this.lockFile)) as GitLockInfo;

      // Validate lock info structure
      if (!lockInfo.pid || !lockInfo.timestamp || !lockInfo.operation) {
        throw new Error('Invalid lock file structure');
      }

      return lockInfo;
    } catch (error) {
      console.error('Error reading lock file:', error);
      return null;
    }
  }

  /**
   * Check if a process is still running
   */
  private isProcessRunning(pid: number): boolean {
    try {
      // On Unix systems, sending signal 0 checks if process exists
      process.kill(pid, 0);
      return true;
    } catch (_error) {
      // Process not found or permission denied
      return false;
    }
  }

  /**
   * Clean up corrupt lock file
   */
  private async cleanupCorruptLock(): Promise<void> {
    try {
      if (await fs.pathExists(this.lockFile)) {
        await fs.remove(this.lockFile);
        console.log('Cleaned up corrupt lock file');
      }
    } catch (error) {
      console.error('Error cleaning up corrupt lock:', error);
    }
  }

  /**
   * Sleep for specified milliseconds
   */
  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}
